### DISCUSSION TITLE
Live candlestick chart

### DISCUSSION THREAD

**Discussion by SwayStar123** [fabianboesiger/ftx#109] [Q&A] [answered]
I am attempting to make my own charting app, im able to fetch historical candles easily using the GetHistoricalPrices endpoint, however rendering the current candle is hard using this, the endpoint has a 3-5 second response time so the live candle would only be updated every 5 seconds at worst if i decide to use that endpoint for this task. I am currently initializing a current candle and then modifying it with new data from the GetTrades endpoint, as that seems to be much faster albeit still too slow. 

How can I make a live candle with similar fast live updates on tradingview while still having accurate data?

**Comment by fabianboesiger** [comment, marked as answer]
Instead of using the `GetTrades` endpoint, try using the WebSocket API provided by this crate, and subscribe to the `Trades` channel. This gives you the same data but with less delay.

**Reply by SwayStar123** [reply]
That solves my live update part of the problem, thanks. Do you think i should stick with GetHistoricalPrices to initialize the candle? Its unreliable during times when a new candle is being formed (eg- the moment the clock goes from 4:59 to 5:00) and sometimes returns nothing and causes a panic in my code, so currently ive just put in a 5 second delay before checking for a new candle. ofc this is not ideal so if you've got any better ideas id appreciate it.

**Reply by fabianboesiger** [reply]
I would do the following:

1. Use GetHistoricalPrices for candles in the past.
2. Use the Trades channel from the WebSocket for the current Candle. That way, you can easily update the current candle.
3. At the same time you switch from fetching data for historical candles to the websocket api for the current candle, issue a GetTrades request to fetch trades that you missed while the current candle was already active. You could use the timestamp of the trades to ensure that you didn't miss any trades:

```
while (earliest timestamp received using WebSocket > latest timestamp received using GetTrades) {
  fetch missing trades using GetTrades
}
```

**Reply by SwayStar123** [reply]
Thank you!! I did not think to use GetTrades in batches, they have a 5000 limit which is what ive been stuck on for a while but i think what you suggested will work.
